### PR TITLE
UPSMON REFACTOR

### PR DESCRIPTION
**Pull request by killerswin2**
## What type of PR is this.
1. [ ] Bug
2. [x] Change
3. [x] Enhancement

### What have you changed and why?
Information:
This was done because start up for UPSMON is slow. Moves all UPSMON stuff to a addon and has base arma functions compile all the code instead of having upsmon do it slowly at mission start. ALL EXECVM WERE REMOVED that referenced UPSMON meaning that we no longer compile every time we a call upsmon. 

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 

********************************************************
Notes:

### CODE REVIEW

**Comment by jaj22** [comment]
Was anything changed in here aside from what's listed in information?

**Comment by killerswin2** [comment]
Besides making a cfgfunction, not really. The only stuff I touched was renaming function global so that they called the right code. I mainly convert that script folder into a addon and removed all the execvm's and replaced them with spawns. I did kill UPSMON weird function compilation code because the bis init_functions does that better then UPSMON 15 year old code.

File: A3A/addons/UPSMON/Functions/INIT/fn_Init_UPSMON.sqf [hunk 1]
```diff
@@ -128,7 +127,7 @@ UPSMON_SUPPORT_WEST_UNITS = [];
 UPSMON_SUPPORT_EAST_UNITS = [];
 UPSMON_SUPPORT_GUER_UNITS = [];
 //tracked units array
-UPSMON_Trackednpcs = [];
+UPSMON_fnc_TRACKednpcs = [];
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/INIT/fn_Init_UPSMON.sqf:130]: Case screwup.

File: A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf [hunk 2]
```diff
@@ -157,19 +157,19 @@ _grp setvariable ["UPSMON_NOAI",_isSoldier];
...
+	_hostility = ["Hostility:",0,_UCthis] call UPSMON_fnc_getArg;
 	_grp setvariable ["UPSMON_GrpHostility",_hostility]
 };
 // create _targerpoint on the roads only (by this group)
 _onroad = if ("ONROAD" in _UCthis) then {true} else {false};
 _grp setvariable ["UPSMON_ONROAD",_onroad];
 
 // Group will not throw smoke
-if ("NOSMOKE" in _UCthis) then {_grp setvariable ["UPSMON_NOSMOKE",true]};
+if ("NOSMOKE" in _UCthis) then {_grp setvariable ["UPSMON_fnc_NOSMOKE",true]};
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf:168]: Not a function.

File: A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf [hunk 3]
```diff
@@ -157,19 +157,19 @@ _grp setvariable ["UPSMON_NOAI",_isSoldier];
...
 
 //Do group share infos ?
 _shareinfos = If ("NOSHARE" in _UCthis) then {false} else {true};
-_grp setvariable ["UPSMON_Shareinfos",_shareinfos];
+_grp setvariable ["UPSMON_fnc_Shareinfos",_shareinfos];
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf:172]: Not a function.

File: A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf [hunk 4]
```diff
@@ -218,13 +218,13 @@ If ("LANDDROP" in _UCthis) then {_grp setvariable ["UPSMON_LANDDROP",true];};
 _nowpType = if ("NOWP" in _UCthis) then {1} else {0};
 _nowpType = if ("NOWP2" in _UCthis) then {2} else {_nowpType};
 _nowpType = if ("NOWP3" in _UCthis) then {3} else {_nowpType};
-_grp setvariable ["UPSMON_NOWP",_nowpType];
+_grp setvariable ["UPSMON_fnc_Nowp",_nowpType];
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf:221]: Not a function.

File: A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf [hunk 5]
```diff
@@ -370,14 +370,14 @@ _grp setformation _formation;
...
-_npc = [_npc,_grp] call UPSMON_getleader;
+_npc = [_npc,_grp] call UPSMON_fnc_getleader;
 if (!alive _npc || !canmove _npc || isplayer _npc ) exitwith {};
 
 _grp setvariable ["UPSMON_GrpStatus","GREEN"];
 _grp setvariable ["UPSMON_GrpMission",_grpmission];
 _grp setvariable ["UPSMON_OrgGrpMission",_grpmission];
 _grp setvariable ["UPSMON_Lastinfos",[[0,0,0],[0,0,0]]];
-_grp setvariable ["UPSMON_NOWP",_nowpType];
+_grp setvariable ["UPSMON_fnc_Nowp",_nowpType];
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/INIT/fn_UPSMON.sqf:380]: Not a function.

File: A3A/addons/UPSMON/Functions/UPSMON_REINFORCEMENT/fn_GetReinfPos.sqf [hunk 6]
```diff
@@ -55,8 +55,8 @@ If (UPSMON_reinforcement || _rfidcalled) then
 				{ 
 					_grp setvariable ["UPSMON_ReinforcementSent",true];
 					_grp setvariable ["UPSMON_Grpmission","REINFORCEMENT"];
-					_grp setvariable ["UPSMON_NOWP",0];
-					[_grp,_fixedtargetPos,"MOVE","COLUMN","FULL","AWARE","YELLOW",1] spawn UPSMON_DocreateWP;
+					_grp setvariable ["UPSMON_fnc_Nowp",0];
```
Comment (jaj22) [inline A3A/addons/UPSMON/Functions/UPSMON_REINFORCEMENT/fn_GetReinfPos.sqf:58]: Not a function.

File: A3A/addons/core/Scripts/UPSMON/Get_pos/UPSMON_pos_init.sqf [hunk 7]
```diff
@@ -1,107 +0,0 @@
-/*
```
Comment (jaj22) [inline A3A/addons/core/Scripts/UPSMON/Get_pos/UPSMON_pos_init.sqf:1]: The documentation from this file should probably go into UPSMON_fnc_pos.

File: A3A/addons/core/functions/CREATE/fn_createAIOutposts.sqf [hunk 8]
```diff
@@ -105,7 +105,7 @@ if ((_frontierX) and (_markerX in outposts)) then
 	{
 		_groupX = createGroup _sideX;
 		_veh = _typeVehX createVehicle (_spawnParameter select 0);
-		_nul=[_veh] execVM QPATHTOFOLDER(scripts\UPSMON\MON_artillery_add.sqf);//TODO need delete UPSMON link
+		_nul=[_veh] spawn UPSMON_fnc_artillery_add;;//TODO need delete UPSMON link
```
Comment (jaj22) [inline A3A/addons/core/functions/CREATE/fn_createAIOutposts.sqf:108]: Bonus semicolon. Don't think it matters but whatever.

File: A3A/addons/core/functions/CREATE/fn_createSDKgarrisonsTemp.sqf [hunk 9]
```diff
@@ -26,14 +26,14 @@ if (_typeX isEqualType "") then {
...
-        _nul=[_veh] execVM QPATHTOFOLDER(scripts\UPSMON\MON_artillery_add.sqf);//TODO need delete UPSMON link
+        _nul=[_veh] spawn UPSMON_fnc_artillery_add;//TODO need delete UPSMON link
         _unit assignAsGunner _veh;
         _unit moveInGunner _veh;
         [_veh, teamPlayer] call A3A_fnc_AIVEHinit;
     };
 
     if (_groups isEqualTo []) then {
-        _nul = [leader _groupX, _markerX, "SAFE","SPAWNED","NOVEH2","NOFOLLOW"] execVM QPATHTOFOLDER(scripts\UPSMON.sqf);//TODO need delete UPSMON link
+        _nul = [leader _groupX, _markerX, "SAFE","SPAWNED","NOVEH2","NOFOLLOW"] spawn UPSMON;//TODO need delete UPSMON link
```
Comment (jaj22) [inline A3A/addons/core/functions/CREATE/fn_createSDKgarrisonsTemp.sqf:36]: should be UPSMON_fnc_UPSMON

**Review by jaj22 — CHANGES_REQUESTED** [review]
Seems legit generally. Marked some minor errors.

Problem will be that it's quite non-obvious whether this stuff breaks (barring RPT errors), as we use it very much as a black box and no-one notices when AI stuff is even more busted than usual. Would still be inclined to merge for 3.1 although I suspect it doesn't make much difference after init. Spawn vs execVM saves a bit, but the unit spawning costs are probably orders of magnitude larger.

**Comment by killerswin2** [comment]
test 1: PR 2548
fn_initServer: 814.8271ms
fn_Init_UPSMON: 806.1371ms

test 2: BASE UNSTABLE
fn_initServer: 929.2622ms
init_UPSMON: 926.1029ms

test 3: PR 2548 RHS
fn_initServer: 1969.1221ms
fn_Init_UPSMON: 1969.0442ms

test 4: BASE UNSTABLE RHS
fn_initServer: 1990.2078ms
init_UPSMON: 1987.0305ms

test 5 PR 2548 RHS after mine fix
fn_initServer: 0.9241ms
fn_Init_UPSMON: 0.8559ms

**Comment by killerswin2** [comment]
I'm going to need to have a test of this on main servers one day, but I think I finally got the init times to a low level.

**Review by Bob-Murphy — APPROVED** [review]
changes look consistent - found no typos
this review does not guarantee functionality

**Review by jaj22 — APPROVED** [review]
Ok, I'm not going to do another 1hr pass on this, and it's conceptually sound. Good enough for testing at least. Remind me to do a garrison functionality check before we release.

**Pull request merged**